syntax = "proto3";
package csce662;
import "google/protobuf/timestamp.proto";
import "google/protobuf/empty.proto";

service SNSService {
  rpc Login(LoginRequest) returns (ClientID) {}
  rpc List(ClientID) returns (ListReply) {}
  rpc Follow(FollowRequest) returns (google.protobuf.Empty) {}
  rpc UnFollow(FollowRequest) returns (google.protobuf.Empty) {}
  rpc Timeline(stream Message) returns (stream Message) {}
}

message ClientID {
  int32 id = 1;
}

message ListReply {
  repeated string all_users = 1;
  repeated string followers = 2;
}

message LoginRequest {
  string username = 1;
  string hashed_password = 2;
}

message FollowRequest {
  int32 id = 1;
  string follower = 2;
}

message Message {
  string username = 1;
  string msg = 2;
  google.protobuf.Timestamp timestamp = 3;
}
